third_party/address_sorting/address_sorting.c

```diff
@@ -0,0 +1,405 @@
+  if (first.source_addr_exists &&
+      grpc_sockaddr_get_family(&first.source_addr) == AF_INET6 &&
+      second.source_addr_exists &&
+      grpc_sockaddr_get_family(&second.source_addr) == AF_INET6) {
+    int first_match_length =
+        ipv6_prefix_match_length((struct sockaddr_in6*)&first.source_addr.addr,
+                                 (struct sockaddr_in6*)&first.dest_addr.addr);
+    int second_match_length =
+        ipv6_prefix_match_length((struct sockaddr_in6*)&second.source_addr.addr,
+                                 (struct sockaddr_in6*)&second.dest_addr.addr);
+    return second_match_length - first_match_length;
+  }
+  return 0;
+}
+
+static int rfc_6724_compare(const void* a, const void* b) {
+  const sortable_address first = *(sortable_address*)a;
+  const sortable_address second = *(sortable_address*)b;
+  int out = 0;
+  if ((out = compare_source_addr_exists(first, second))) {
+    return out;
+  } else if ((out = compre_source_dest_scope_matches(first, second))) {
+    return out;
+  } else if ((out = compare_source_dest_labels_match(first, second))) {
+    return out;
+    // TODO: avoid deprecated addresses
+    // TODO: avoid temporary addresses
+  } else if ((out = compare_dest_precedence(first, second))) {
+    return out;
+    // TODO: prefer native transport
+  } else if ((out = compare_dest_scope(first, second))) {
+    return out;
+  } else if ((out = compare_source_dest_prefix_match_lengths(first, second))) {
+    return out;
+  }
+  // Prefer that the sort be stable otherwise
+  return (int)(first.original_index - second.original_index);
+}
+
+void address_sorting_override_socket_factory_for_testing(
+    address_sorting_socket_factory* factory) {
+  GPR_ASSERT(g_current_socket_factory);
+  g_current_socket_factory->vtable->destroy(g_current_socket_factory);
+  g_current_socket_factory = factory;
+}
+
+void address_sorting_rfc_6724_sort(grpc_lb_addresses* resolved_lb_addrs) {
+  sortable_address* sortable =
+      gpr_zalloc(sizeof(sortable_address) * resolved_lb_addrs->num_addresses);
+  for (size_t i = 0; i < resolved_lb_addrs->num_addresses; i++) {
+    sortable_address next;
+    memset(&next, 0, sizeof(sortable_address));
+    next.original_index = i;
+    next.lb_addr = resolved_lb_addrs->addresses[i];
+    next.dest_addr = resolved_lb_addrs->addresses[i].address;
+    int address_family =
+        grpc_sockaddr_get_family(&resolved_lb_addrs->addresses[i].address);
+    int s =
+        address_sorting_socket(address_family, SOCK_DGRAM | SOCK_CLOEXEC, 0);
+    if (s != -1) {
+      if (address_sorting_connect(
+              s,
+              (struct sockaddr*)&resolved_lb_addrs->addresses[i].address.addr,
+              (socklen_t)resolved_lb_addrs->addresses[i].address.len) != -1) {
+        grpc_resolved_address found_source_addr;
+        if (address_sorting_getsockname(
+                s, (struct sockaddr*)&found_source_addr.addr,
+                (socklen_t*)&found_source_addr.len) != -1) {
```

Comment:
`found_source_addr.len` must be initialized.

Reply:
fixed